sunpy/coordinates/transformations.py

```diff
@@ -76,108 +63,155 @@ def _observers_are_equal(obs_1, obs_2, string_ok=False):
+                                 HeliographicCarrington, HeliographicStonyhurst)
 def hgc_to_hgs(hgccoord, hgsframe):
     """
-    Convert from Heliograpic Carrington to Heliographic Stonyhurst.
+    Convert from Heliographic Carrington to Heliographic Stonyhurst.
     """
-    if hgsframe.obstime is None or np.any(hgsframe.obstime != hgccoord.obstime):
-        raise ValueError("Can not transform from Heliographic Carrington to "
-                         "Heliographic Stonyhurst, unless both frames have matching obstime.")
-    obstime = hgsframe.obstime
-    s_lon = hgccoord.spherical.lon - _carrington_offset(obstime).to(
-        u.deg)
-    representation = SphericalRepresentation(s_lon, hgccoord.spherical.lat,
-                                             hgccoord.spherical.distance)
+    # First transform the HGC coord to the HGS obstime
+    int_coord = _transform_obstime(hgccoord, hgsframe.obstime)
+
+    # Rotate from HGC to HGS
+    total_matrix = matrix_transpose(_rotation_matrix_hgs_to_hgc(int_coord.obstime))
+    newrepr = int_coord.cartesian.transform(total_matrix)
+
+    return hgsframe.realize_frame(newrepr)
 
-    return hgsframe.realize_frame(representation)
 
+def _matrix_hcc_to_hpc():
+    # Returns the transformation matrix that permutes/swaps axes from HCC to HPC
 
-@frame_transform_graph.transform(FunctionTransform, Heliocentric,
-                                 Helioprojective)
+    # HPC spherical coordinates are a left-handed frame with these equivalent Cartesian axes:
+    #   HPC_X = -HCC_Z
+    #   HPC_Y = HCC_X
+    #   HPC_Z = HCC_Y
+    # (HPC_X and HPC_Y are not to be confused with HPC_Tx and HPC_Ty)
+    return np.array([[0, 0, -1],
+                     [1, 0, 0],
+                     [0, 1, 0]])
+
+
+@frame_transform_graph.transform(FunctionTransformWithFiniteDifference,
+                                 Heliocentric, Helioprojective)
 def hcc_to_hpc(helioccoord, heliopframe):
     """
-    Convert from Heliocentic Cartesian to Helioprojective Cartesian.
+    Convert from Heliocentric Cartesian to Helioprojective Cartesian.
     """
-    if not _observers_are_equal(helioccoord.observer, heliopframe.observer):
-        heliocframe = Heliocentric(observer=heliopframe.observer)
-        new_helioccoord = helioccoord.transform_to(heliocframe)
-        helioccoord = new_helioccoord
-
-    x = helioccoord.x.to(u.m)
-    y = helioccoord.y.to(u.m)
-    z = helioccoord.z.to(u.m)
+    # Loopback transform HCC coord to obstime and observer of HPC frame
+    int_frame = Heliocentric(obstime=heliopframe.obstime, observer=heliopframe.observer)
+    int_coord = helioccoord.transform_to(int_frame)
 
-    # d is calculated as the distance between the points
-    # (x,y,z) and (0,0,D0).
-    distance = np.sqrt(x**2 + y**2 + (helioccoord.observer.radius - z)**2)
+    # Shift the origin from the Sun to the observer
+    distance = SkyCoord(int_coord.observer).heliographic_stonyhurst.radius
```

Comment:
I don't think we want to incur the performance hit of the `SkyCoord` initializer here, would be better to do:
```suggestion
    distance = int_coord.observer.transform_to("heliographic_stonyhurst").radius
```

Although, because `int_coord` is always an instance of `Heliocentric` we actually know that `.observer` will always be a HGS frame, because the Descriptor for the frame always returns it in that frame irrespective of what was passed in.

Reply:
Using `transform_to("heliographic_stonyhurst")` would actually be a bit problematic in some cases, but if it's guaranteed to be HGS, then I'll just take it out.

Reply:
Isn't that what `.heliographic_stonyhurst` is doing underneath anyway?